cmd/erasure-zones.go

```diff
@@ -541,22 +542,33 @@ func (z *erasureZones) GetObjectInfo(ctx context.Context, bucket, object string,
 
 // PutObject - writes an object to least used erasure zone.
 func (z *erasureZones) PutObject(ctx context.Context, bucket string, object string, data *PutObjReader, opts ObjectOptions) (ObjectInfo, error) {
-	// Lock the object.
-	lk := z.NewNSLock(ctx, bucket, object)
-	if err := lk.GetLock(globalObjectTimeout); err != nil {
-		return ObjectInfo{}, err
-	}
-	defer lk.Unlock()
+	// Get lock for the object.
+	data.lock = z.NewNSLock(ctx, bucket, object)
 
 	if z.SingleZone() {
 		return z.zones[0].PutObject(ctx, bucket, object, data, opts)
 	}
 
-	idx, err := z.getZoneIdx(ctx, bucket, object, opts, data.Size())
+	// Get a lock until we have determined a zone.
+	if err := data.lock.GetLock(globalObjectTimeout); err != nil {
+		return ObjectInfo{}, err
+	}
+
+	idx, fixed, err := z.getZoneIdx(ctx, bucket, object, opts, data.Size())
 	if err != nil {
+		data.lock.Unlock()
 		return ObjectInfo{}, err
 	}
 
+	if fixed {
+		// We don't need to keep the lock since zone is determined.
+		data.lock.Unlock()
```

Comment:
@klauspost this can lead to races - the current code is implemented on purpose since we hold global lock per object. So any unnecessary concurrent operation on the object is avoided. 

We cannot avoid these races and they are quite common - leaving room for any such occurrence is more disastrous than blocking.

Listing blocked original issue is not possible since we don't hold any locks in Listing. Unless the application is issuing GetObjectInfo() on the object where it can get blocked.